### PR TITLE
gh-19, gh-20, gh-21 Add Task Tab to UI

### PR DESCRIPTION
**Pull request by ghillert**
* Migrate similar code from the Jobs Tab code
* Add support to view Task Modules
* Add support to create Task Definitions
* Add support to launch Tasks
* Add support to view Task Executions
* Remove `job` references from Task code
* Remove unnecessary code under the Job module
* Add additional check-box when creating Task definitions - Only when checked will properties be included in the definition
* Optimize Task Launch screen - remove support for `typed` parameters. Tasks will always be String values.
* Fix: When launching tasks, properties need to be passed as comma separated key-value string instead of JSON String to REST endpoint

See:

* https://github.com/spring-cloud/spring-cloud-dataflow-ui/issues/19
* https://github.com/spring-cloud/spring-cloud-dataflow-ui/issues/20
* https://github.com/spring-cloud/spring-cloud-dataflow-ui/issues/21

Keep in mind of the following issues:

* The task executions screen will have date formatting issues due to https://github.com/spring-cloud/spring-cloud-dataflow/issues/434
* Depends on https://github.com/spring-projects/spring-boot/pull/5335

Resolves spring-cloud/spring-cloud-dataflow-ui#19
Resolves spring-cloud/spring-cloud-dataflow-ui#20
Resolves spring-cloud/spring-cloud-dataflow-ui#21

### CODE REVIEW

File: ui/app/scripts/task/views/executions.html [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+			<th>Execution Id</th><th>Task Name</th><th>Start Time</th>
+			<th>End Time</th><th>Exit Code</th>
+		</tr>
+	</thead>
+	<tbody>
+		<tr dir-paginate="item in pageable.items | itemsPerPage: pageable.pageSize | filter:filterQuery" total-items="pageable.total"
+		    ng-show="!item.inactive">
+			<td>{{item.executionId}}</td>
+			<td>{{item.taskName}}</td>
+			<td xd-date-time="{{item.startTime}}"></td>
```
Comment (ilayaperumalg) [inline ui/app/scripts/task/views/executions.html]: This `startTime` value seems to have broken when used with `xd-date-time` directive. I see `invalid Date` as the value.

File: ui/app/scripts/task/views/executions.html [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+			<td xd-date-time="{{item.endTime}}"></td>
+			<td>{{item.exitCode}}</td>
```
Comment (ilayaperumalg) [inline ui/app/scripts/task/views/executions.html]: Are the `step execution count` and the `Actions` to view Job Execution details that would lead to step execution detail etc., not part of this PR?

File: ui/app/scripts/task/views/executions.html (hunk 1 above)
Comment (ilayaperumalg) [inline ui/app/scripts/task/views/executions.html]: I see this is a know issue as you have already created https://github.com/spring-cloud/spring-cloud-dataflow/issues/434 for the date formatting issue.

File: ui/app/scripts/task/views/definitions.html [hunk 3]
```diff
@@ -21,21 +21,12 @@
   <tbody>
     <tr dir-paginate="item in pageable.items | itemsPerPage: pageable.pageSize | filter:filterQuery" total-items="pageable.total" ng-show="!item.inactive">
       <td>{{item.name}}</td>
-      <td xd-format-stream='item.definition'></td>
+      <td xd-format-stream='item.dslText'></td>
       <td>{{item.status}}</td>
```
Comment (ilayaperumalg) [inline ui/app/scripts/task/views/definitions.html]: Not part of this PR, but is it normal to see the status of the task as `deploying` always though the task app says it started? cc @mminella @cppwfs

File: ui/app/scripts/task/views/executions.html (hunk 2 above)
Comment (ilayaperumalg) [inline ui/app/scripts/task/views/executions.html]: Given we have a story https://github.com/spring-cloud/spring-cloud-dataflow-ui/issues/22 to track executions tab, may be the entire page implementation is the scope of that story.

File: ui/app/scripts/routes.js [hunk 4]
```diff
@@ -77,13 +85,21 @@ define(['./app'], function (xdAdmin) {
       }
     })
     .state('home.jobs.tabs', {
-      url : 'tasks',
+      url : 'jobs',
```
Comment (ilayaperumalg) [inline ui/app/scripts/routes.js]: Why `jobs` here? Also, I am confused by the existence of `job` based routes (vs `task` based ones) in this file.

Comment (ilayaperumalg) [inline ui/app/scripts/routes.js]: I think we can get rid of all these `job` based routes.

File: ui/app/index.html [hunk 5]
```diff
@@ -46,8 +46,11 @@
             <li ng-class="{ active: $state.includes('home.streams') }">
                <a ui-sref="home.streams.tabs.definitions">Streams</a>
             </li>
+            <li ng-class="{ active: $state.includes('home.tasks') }">
+              <a ui-sref="home.tasks.tabs.definitions">Tasks</a>
+            </li>
             <li ng-class="{ active: $state.includes('home.jobs') }">
```
Comment (ilayaperumalg) [inline ui/app/index.html]: What is the reason behind keeping the `jobs` tab with the **executions** page in it? I thought we are going to have the detailed `executions` page under the `tasks -> executions` anyway and this tab can be removed?

File: ui/app/scripts/app.js [hunk 6]
```diff
@@ -23,6 +23,7 @@ define([
   'angular',
   './analytics/app',
   './job/app',
```
Comment (ilayaperumalg) [inline ui/app/scripts/app.js]: Unless there is a specific reason, we can complete remove the mappings of `job` and associated resources (scripts, routes etc.,)?

**Comment by ilayaperumalg** [comment]
Created https://github.com/spring-cloud/spring-cloud-dataflow-ui/issues/31 and https://github.com/spring-cloud/spring-cloud-dataflow-ui/issues/32 to remove the XD references and renaming `admin` -> `server` in general.

File: ui/app/scripts/task/controllers/executions.js [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+      utils.addBusyPromise(taskExcutionsPromise);
+
+      taskExcutionsPromise.then(
+          function (result) {
+            utils.$log.info('task excutions', result);
+            $scope.pageable.items = result._embedded.taskExecutionResourceList;
+            $scope.pageable.total = result.page.totalElements;
+            utils.$log.info('$scope.pageable', $scope.pageable);
+          }, function () {
+            utils.growl.error('Error fetching data. Is the XD server running?');
```
Comment (ilayaperumalg) [inline ui/app/scripts/task/controllers/executions.js]: No XD server ;)

File: ui/app/index.html (hunk 5 above)
Comment (ghillert) [inline ui/app/index.html]: @ilayaperumalg @mminella  @cppwfs  Nope - The Jobs tab will contain Job-specific info. In this case Job Executions. Eventually a Job Execution should link back to the corresponding Task Execution. 

A Task-Execution **may** or **may-not** have a corresponding Job Execution but a Job Execution will always have a Task Execution. 

Long-term, we may even re-add more features back to the Jobs tab. There is even the UX team working on this right now to provide more guidance.

File: ui/app/scripts/app.js (hunk 6 above)
Comment (ghillert) [inline ui/app/scripts/app.js]: As stated above, we will need the jobs tab.

File: ui/app/scripts/task/controllers/executions.js (hunk 7 above)
Comment (ghillert) [inline ui/app/scripts/task/controllers/executions.js]: Having that as a separate PR would be good. There are just too many XD references.

File: ui/app/scripts/routes.js (hunk 4 above)
Comment (ghillert) [inline ui/app/scripts/routes.js]: See my other comment below.

File: ui/app/scripts/task/views/definitions.html (hunk 3 above)
Comment (ghillert) [inline ui/app/scripts/task/views/definitions.html]: Will look at that - might be a UI issue.

File: ui/app/scripts/task/views/executions.html (hunk 2 above)
Comment (ghillert) [inline ui/app/scripts/task/views/executions.html]: Task Executions is different from Job Executions. Task Executions are higher-level and more generic. It should be executing a shell-script aka does not necessitate the execution of a batch job.

File: ui/app/scripts/task/views/executions.html (hunk 1 above)
Comment (ghillert) [inline ui/app/scripts/task/views/executions.html]: Yes - I did not want to change the directive - I strongly believe Json response should send ISO formatted dates.

File: ui/app/scripts/task/views/definitions.html (hunk 3 above)
Comment (ghillert) [inline ui/app/scripts/task/views/definitions.html]: The state is coming from the server.

```
		@Override
		public TaskDefinitionResource instantiateResource(TaskDefinition taskDefinition) {
			ModuleDeploymentId id =
					ModuleDeploymentId.fromModuleDefinition(taskDefinition.getModuleDefinition());
			TaskDefinitionResource taskDefinitionResource = new TaskDefinitionResource(taskDefinition.getName(),
					taskDefinition.getDslText());
			taskDefinitionResource.setStatus(moduleDeployer.status(id).getState().name());
			return taskDefinitionResource;
		}
```

**Comment by ilayaperumalg** [comment]
LGTM, @ghillert could you merge this PR please?

**Comment by ghillert** [comment]
Merged with 3217fb2e7b02e28953df611f672dc05db03185db
